### PR TITLE
Support HTML parser reentrancy

### LINKED ISSUE #24115: Borrow hazard when calling Tokenizer::end from ServoParser::do_parse_sync
With zealous GC, we expose a borrow failure when ending tokenizing:
```
 4:16.33 pid:93228 DomRefCell<T> already mutably borrowed: BorrowError (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at src/libcore/result.rs:1084)
 4:17.00 pid:93228 stack backtrace:
 4:17.00 pid:93228    0: __ZN9backtrace9backtrace5trace17hc4a0d1f273e07024E (0x10f073d7b)
 4:17.00 pid:93228    1: __ZN9backtrace7capture9Backtrace3new17h765760e39b281a50E (0x10f072f67)
 4:17.00 pid:93228    2: __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h510947304905906eE (0x10b7d47db)
 4:17.00 pid:93228    3: __ZN3std9panicking20rust_panic_with_hook17h099fe3773ce1a793E (0x10f1ffe41)
 4:17.00 pid:93228    4: __ZN3std9panicking18continue_panic_fmt17h8a4b76634870b383E (0x10f1ff8ac)
 4:17.00 pid:93228    5: _rust_begin_unwind (0x10f1ff7a8)
 4:17.00 pid:93228    6: __ZN4core9panicking9panic_fmt17h885799c19d2b8b0cE (0x10f21ffde)
 4:17.01 pid:93228    7: __ZN4core6result13unwrap_failed17h91161a3faccc6622E (0x10f2200b8)
 4:17.01 pid:93228    8: __ZN110_$LT$script..dom..bindings..cell..DomRefCell$LT$T$GT$$u20$as$u20$script..dom..bindings..trace..JSTraceable$GT$5trace17had7eda3ca734e750E (0x10bdfd3d8)
 4:17.01 pid:93228    9: ___rust_maybe_catch_panic (0x10f2081de)
 4:17.01 pid:93228   10: __ZN6script3dom8bindings7codegen8Bindings18ServoParserBinding18ServoParserBinding6_trace17hcb09c9acba2f1213E (0x10c512fb7)
 4:17.01 pid:93228   11: __ZNK2js5Class7doTraceEP8JSTracerP8JSObject (0x10d1cebdb)
 4:17.01 pid:93228   12: __ZL13CallTraceHookIZN2js8GCMarker19processMarkStackTopERNS0_11SliceBudgetEEUlT_E_EPNS0_12NativeObjectEOS4_P8JSTracerP8JSObject15CheckGeneration (0x10d45f583)
 4:17.01 pid:93228   13: __ZN2js8GCMarker19processMarkStackTopERNS_11SliceBudgetE (0x10d417ecf)
 4:17.01 pid:93228   14: __ZN2js8GCMarker24markUntilBudgetExhaustedERNS_11SliceBudgetE (0x10d3fc134)
 4:17.01 pid:93228   15: __ZN2js2gc9GCRuntime24markUntilBudgetExhaustedERNS_11SliceBudgetENS_7gcstats9PhaseKindE (0x10d3f8ee5)
 4:17.01 pid:93228   16: __ZN2js2gc9GCRuntime16incrementalSliceERNS_11SliceBudgetEN2JS8GCReasonERNS0_13AutoGCSessionE (0x10d401163)
 4:17.02 pid:93228   17: __ZN2js2gc9GCRuntime7gcCycleEbNS_11SliceBudgetEN2JS8GCReasonE (0x10d402e32)
 4:17.02 pid:93228   18: __ZN2js2gc9GCRuntime7collectEbNS_11SliceBudgetEN2JS8GCReasonE (0x10d4044e3)
 4:17.02 pid:93228   19: __ZN2js2gc9GCRuntime2gcE18JSGCInvocationKindN2JS8GCReasonE (0x10d3cdcb6)
 4:17.02 pid:93228   20: __ZN2js2gc9GCRuntime10runDebugGCEv (0x10d3cda34)
 4:17.02 pid:93228   21: __ZN2js2gc9GCRuntime22gcIfNeededAtAllocationEP9JSContext (0x10d3cd3a5)
 4:17.02 pid:93228   22: __ZN2js2gc9GCRuntime19checkAllocatorStateILNS_7AllowGCE1EEEbP9JSContextNS0_9AllocKindE (0x10d3c89f1)
 4:17.02 pid:93228   23: __ZN2js8AllocateINS_11ObjectGroupELNS_7AllowGCE1EEEPT_P9JSContext (0x10d3cb918)
 4:17.02 pid:93228   24: __ZN2js16ObjectGroupRealm9makeGroupEP9JSContextPN2JS5RealmEPKNS_5ClassENS3_6HandleINS_11TaggedProtoEEEj (0x10ce86688)
 4:17.02 pid:93228   25: __ZN8JSObject13makeLazyGroupEP9JSContextN2JS6HandleIPS_EE (0x10ce8646b)
 4:17.02 pid:93228   26: __ZN8JSObject8getGroupEP9JSContextN2JS6HandleIPS_EE (0x10cdf634e)
 4:17.02 pid:93228   27: __ZN8JSObject15splicePrototypeEP9JSContextN2JS6HandleIPS_EENS3_IN2js11TaggedProtoEEE (0x10ce85f09)
 4:17.02 pid:93228   28: __Z18JS_SplicePrototypeP9JSContextN2JS6HandleIP8JSObjectEES5_ (0x10d210b9b)
 4:17.02 pid:93228   29: __Z26JS_NewObjectWithUniqueTypeP9JSContextPK7JSClassN2JS6HandleIP8JSObjectEE (0x10d210d5b)
 4:17.02 pid:93228   30: __ZN6script3dom8bindings9interface13create_object17h8986b5acf9af4f2dE (0x10c61d1b5)
 4:17.02 pid:93228   31: __ZN6script3dom8bindings9interface33create_interface_prototype_object17h88cc524928e422e3E (0x10c61c9ee)
 4:17.02 pid:93228   32: __ZN6script3dom8bindings7codegen8Bindings22HTMLBodyElementBinding22HTMLBodyElementBinding22CreateInterfaceObjects17h2b8b8414feb603deE.llvm.4880966149733538817 (0x10c860ebe)
 4:17.02 pid:93228   33: __ZN6script3dom8bindings7codegen8Bindings22HTMLBodyElementBinding22HTMLBodyElementBinding4Wrap17he01419c0c8076f7fE (0x10c860839)
 4:17.02 pid:93228   34: __ZN6script3dom6create26create_native_html_element17h5f0ccfb8d68fb478E (0x10bf71e84)
 4:17.02 pid:93228   35: __ZN6script3dom6create14create_element17h16d5824be611a653E (0x10bf73c2b)
 4:17.02 pid:93228   36: __ZN6script3dom11servoparser24create_element_for_token17h26510e09c0520b4fE (0x10c2ffc65)
 4:17.02 pid:93228   37: __ZN97_$LT$script..dom..servoparser..Sink$u20$as$u20$markup5ever..interface..tree_builder..TreeSink$GT$14create_element17hc33c3c0aeb0eb546E (0x10c2fcb7d)
 4:17.02 pid:93228   38: __ZN11markup5ever9interface12tree_builder14create_element17h90bb47d0a1bca068E (0x10bf99153)
 4:17.03 pid:93228   39: __ZN9html5ever12tree_builder32TreeBuilder$LT$Handle$C$Sink$GT$14insert_element17hb678b927dc5a81cdE (0x10c7dd964)
 4:17.03 pid:93228   40: __ZN9html5ever12tree_builder32TreeBuilder$LT$Handle$C$Sink$GT$4step17he4f4df0d0604f86bE (0x10c7f4e14)
 4:17.03 pid:93228   41: __ZN120_$LT$html5ever..tree_builder..TreeBuilder$LT$Handle$C$Sink$GT$$u20$as$u20$html5ever..tokenizer..interface..TokenSink$GT$13process_token17h6adb751266517276E (0x10c78ca34)
 4:17.03 pid:93228   42: __ZN9html5ever9tokenizer21Tokenizer$LT$Sink$GT$13process_token17hd3e0905ceb9fdbefE.llvm.13403837931426071822 (0x10beb7802)
 4:17.03 pid:93228   43: __ZN6script3dom11servoparser4html9Tokenizer3end17hbf1b2123742649a0E (0x10beeb28a)
 4:17.03 pid:93228   44: __ZN6script3dom11servoparser11ServoParser13do_parse_sync17ha608c3845fd71bd5E (0x10c2f67c0)
 4:17.03 pid:93228   45: __ZN14profile_traits4time7profile17h5030af151774cb74E (0x10c5f603d)
 4:17.03 pid:93228   46: __ZN6script3dom11servoparser11ServoParser10parse_sync17h5e54198daa0f3441E.llvm.15491451777926491180 (0x10c2f54da)
 4:17.03 pid:93228   47: __ZN6script3dom11servoparser11ServoParser43resume_with_pending_parsing_blocking_script17h84c2e7bcb8c48ce6E (0x10c2f33e2)
 4:17.03 pid:93228   48: __ZN6script3dom8document8Document39process_pending_parsing_blocking_script17h19a631d52f765e5cE (0x10c44cafd)
 4:17.03 pid:93228   49: __ZN6script3dom8document8Document38pending_parsing_blocking_script_loaded17he380e1268a6a6f9cE (0x10c44c6ae)
 4:17.03 pid:93228   50: __ZN99_$LT$script..dom..htmlscriptelement..ScriptContext$u20$as$u20$net_traits..FetchResponseListener$GT$20process_response_eof17hbeea62ba1532241eE (0x10c59c1e5)
 4:17.03 pid:93228   51: __ZN76_$LT$net_traits..FetchResponseMsg$u20$as$u20$net_traits..Action$LT$T$GT$$GT$7process17h1e3f419940403dd4E (0x10c8c7e67)
 4:17.03 pid:93228   52: __ZN101_$LT$script..network_listener..ListenerTask$LT$A$C$Listener$GT$$u20$as$u20$script..task..TaskOnce$GT$8run_once17h53b608f73558fd98E (0x10c75bdac)
 4:17.03 pid:93228   53: __ZN43_$LT$T$u20$as$u20$script..task..TaskBox$GT$7run_box17hdd2d15a7c0a73618E (0x10c8f159e)
 4:17.03 pid:93228   54: __ZN6script13script_thread12ScriptThread22handle_msg_from_script17h9e03eb14b1d6d27aE (0x10c4c8cf7)
 4:17.03 pid:93228   55: __ZN6script13script_thread12ScriptThread11handle_msgs17h689a61696977da24E.llvm.16624544286892866357 (0x10c4c0b1b)
 4:17.03 pid:93228   56: __ZN14profile_traits3mem12ProfilerChan25run_with_memory_reporting17h7b6e9bd38e1ab8feE (0x10c5f3c87)
 4:17.03 pid:93228   57: __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17he46720d101a3e6efE (0x10c8a10bb)
 4:17.04 pid:93228   58: __ZN3std9panicking3try7do_call17h5f1d988e7dc73db4E.llvm.16932765531113996662 (0x10bd6c54a)
 4:17.04 pid:93228   59: ___rust_maybe_catch_panic (0x10f2081de)
 4:17.04 pid:93228   60: __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17hc92a2538c3cac723E (0x10bfa8755)
 4:17.04 pid:93228   61: __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17hb03304ace7f89c63E (0x10f1ef56d)
 4:17.04 pid:93228   62: __ZN3std3sys4unix6thread6Thread3new12thread_start17hd2d570f25bb299f3E (0x10f20770d)
 4:17.04 pid:93228   63: __pthread_body (0x7fff63578660)
 4:17.04 pid:93228   64: __pthread_start (0x7fff6357850c)
```

### LINKED ISSUE #32454: HTML parser does not support reentrancy via custom elements
**Describe the bug:**
I've been trying to access my website ([kilonova.ro](https://kilonova.ro)) using Servo ever since I heard of the project. However, every time I get greeted with the following failed assertion (Console output with `RUST_BACKTRAGE=1`):
```rs
assertion failed: window.Document().can_invoke_script() (thread Script(1,1), at components/script/dom/eventtarget.rs:400)
0: servoshell::backtrace::print
1: servoshell::main::{{closure}}
2: std::panicking::rust_panic_with_hook
3: std::panicking::begin_panic_handler::{{closure}}
4: std::sys_common::backtrace::__rust_end_short_backtrace
5: rust_begin_unwind
6: core::panicking::panic_fmt
7: core::panicking::panic
8: script::dom::eventtarget::EventTarget::dispatch_event
9: script::dom::bindings::codegen::Bindings::EventTargetBinding::EventTarget_Binding::dispatchEvent::{{closure}}
10: mozjs::panic::wrap_panic
11: script::dom::bindings::codegen::Bindings::EventTargetBinding::EventTarget_Binding::dispatchEvent
12: CallJitMethodOp
13: script::dom::bindings::utils::generic_call
14: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
15: _ZN2js9InterpretEP9JSContextRNS_8RunStateE
16: _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
17: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
18: _ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
19: _Z20JS_CallFunctionValueP9JSContextN2JS6HandleIP8JSObjectEENS2_INS1_5ValueEEERKNS1_16HandleValueArrayENS1_13MutableHandleIS6_EE
20: script::dom::bindings::codegen::Bindings::FunctionBinding::Function::Call
21: script::dom::customelementregistry::CustomElementReaction::invoke
22: script::dom::element::Element::invoke_reactions
23: script::dom::customelementregistry::ElementQueue::invoke_reactions
24: script::dom::customelementregistry::CustomElementReactionStack::pop_current_element_queue
25: html5ever::tree_builder::TreeBuilder<Handle,Sink>::insert_element
26: html5ever::tree_builder::TreeBuilder<Handle,Sink>::step
27: <html5ever::tree_builder::TreeBuilder<Handle,Sink> as html5ever::tokenizer::interface::TokenSink>::process_token
28: html5ever::tokenizer::Tokenizer<Sink>::process_token
29: html5ever::tokenizer::Tokenizer<Sink>::emit_current_tag
30: html5ever::tokenizer::Tokenizer<Sink>::step
31: html5ever::tokenizer::Tokenizer<Sink>::run
32: script::dom::servoparser::Tokenizer::feed
33: script::dom::servoparser::ServoParser::do_parse_sync
34: profile_traits::time::profile
35: script::dom::servoparser::ServoParser::parse_sync
36: script::dom::document::Document::process_pending_parsing_blocking_script
37: script::dom::document::Document::finish_load
38: <script::stylesheet_loader::StylesheetContext as net_traits::FetchResponseListener>::process_response_eof
39: <net_traits::FetchResponseMsg as net_traits::Action<T>>::process
40: <script::network_listener::ListenerTask<A,Listener> as script::task::TaskOnce>::run_once
41: <T as script::task::TaskBox>::run_box
42: script::script_thread::ScriptThread::handle_msg_from_script
43: script::script_thread::ScriptThread::handle_msgs
44: profile_traits::mem::ProfilerChan::run_with_memory_reporting
45: std::sys_common::backtrace::__rust_begin_short_backtrace
46: core::ops::function::FnOnce::call_once{{vtable.shim}}
47: std::sys::pal::unix::thread::Thread::new::thread_start
48: start_thread
49: __GI___clone3
[2024-06-07T09:36:37Z ERROR servoshell] assertion failed: window.Document().can_invoke_script()
```

**To Reproduce:**
- Start Servo nightly
- Go to https://kilonova.ro/
- See the error

My theory is that it must be something regarding Web Components created by preact. Here's a more isolated HTML document that shows the problem clearer:

```html
<!DOCTYPE html>
<html>
	<head>
		<script>
			// Boilerplate the platform's bundle expects to exist, not important
			window.platform_info = { language: "en" }
		</script>
		<!-- The platform's javascript bundle. A map file is generated alongside it. It's generated using esbuild using files from https://github.com/KiloProjects/Kilonova/tree/master/web/assets -->
		<script src="https://kilonova.ro/static/bundle.js"></script>
	</head>
	<body>
		<h1>Reproduction example</h1>
		<!-- Commenting the following line makes Servo execute normally. If uncommented, Servo crashes. -->
		<kn-glossary name="stdin" content="Hello"></kn-glossary>
		
		<!-- Boilerplate the platform's bundle expects to exist, not important -->
		<div id="modals"></div>
	</body>
</html>
```

I have tried to isolate the bug even more and make a more reproducible example but failed to do so.

**Platform:**
OS: Fedora Linux 40 (but Fedora 39 also didn't work)
Hardware: x86_64 Ryzen 5 5600x, but I don't think it's really relevant here since it seems like a logic/JS issue

### LINKED ISSUE #32646: Intermittent double mutable borrow when GC occurs during synchronous parsing
From interpreting the stack trace, this looks like the same underlying issue as https://github.com/servo/servo/issues/32454. When the parser is synchronously parsing and a GC occurs, we attempt to mutably borrow fields of the parser that are already borrowed.

<details>

```
  ▶ CRASH [expected OK] /url/failure.html (#28574)
  │ 
  │ already mutably borrowed: BorrowError (thread Script(1,1), at components/script/dom/bindings/cell.rs:80)
  │    0: servoshell::backtrace::print
  │    1: servoshell::panic_hook::panic_hook
  │    2: std::panicking::rust_panic_with_hook
  │    3: std::panicking::begin_panic_handler::{{closure}}
  │    4: std::sys_common::backtrace::__rust_end_short_backtrace
  │    5: rust_begin_unwind
  │    6: core::panicking::panic_fmt
  │    7: core::cell::panic_already_mutably_borrowed
  │    8: script::dom::bindings::trace::<impl mozjs::gc::trace::Traceable for script::dom::bindings::cell::DomRefCell<T>>::trace
  │    9: script::dom::bindings::codegen::Bindings::ServoParserBinding::ServoParser_Binding::_trace::{{closure}}
  │   10: mozjs::panic::wrap_panic
  │   11: script::dom::bindings::codegen::Bindings::ServoParserBinding::ServoParser_Binding::_trace
  │   12: _ZN2js8GCMarker19processMarkStackTopILj4EEEbRNS_11SliceBudgetE
  │   13: _ZN2js8GCMarker9doMarkingILj4EEEbRNS_11SliceBudgetENS_2gc20ShouldReportMarkTimeE
  │   14: _ZN2js2gc9GCRuntime24markUntilBudgetExhaustedERNS_11SliceBudgetENS1_15ParallelMarkingENS0_20ShouldReportMarkTimeE
  │   15: _ZN2js2gc9GCRuntime16incrementalSliceERNS_11SliceBudgetEN2JS8GCReasonEb
  │   16: _ZN2js2gc9GCRuntime7gcCycleEbRKNS_11SliceBudgetEN2JS8GCReasonE
  │   17: _ZN2js2gc9GCRuntime7collectEbRKNS_11SliceBudgetEN2JS8GCReasonE
  │   18: _ZN2js2gc9GCRuntime17gcIfRequestedImplEb
  │   19: _ZN2js2gc13CellAllocator16AllocTenuredCellILNS_7AllowGCE1EEEPvP9JSContextNS0_9AllocKindEm
  │   20: _ZN2js12GetterSetter6createEP9JSContextN2JS6HandleIP8JSObjectEES7_
  │   21: _ZN2js20NativeDefinePropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_INS2_11PropertyKeyEEENS3_INS2_18PropertyDescriptorEEERNS2_14ObjectOpResultE
  │   22: _ZN2js22DefineAccessorPropertyEP9JSContextN2JS6HandleIP8JSObjectEENS3_INS2_11PropertyKeyEEES6_S6_j
  │   23: _ZL26DefineAccessorPropertyByIdP9JSContextN2JS6HandleIP8JSObjectEENS2_INS1_11PropertyKeyEEERK15JSNativeWrapperSA_j
  │   24: _Z19JS_DefinePropertiesP9JSContextN2JS6HandleIP8JSObjectEEPK14JSPropertySpec
  │   25: mozjs::rust::define_properties
  │   26: script::dom::bindings::interface::create_object
  │   27: script::dom::bindings::interface::create_interface_prototype_object
  │   28: script::dom::bindings::codegen::Bindings::HTMLBodyElementBinding::HTMLBodyElement_Binding::CreateInterfaceObjects
  │   29: script::dom::bindings::interface::get_per_interface_object_handle
  │   30: script::dom::bindings::codegen::Bindings::HTMLBodyElementBinding::HTMLBodyElement_Binding::Wrap
  │   31: script::dom::create::create_native_html_element
  │   32: script::dom::create::create_element
  │   33: script::dom::servoparser::create_element_for_token
  │   34: <script::dom::servoparser::Sink as markup5ever::interface::tree_builder::TreeSink>::create_element
  │   35: markup5ever::interface::tree_builder::create_element
  │   36: html5ever::tree_builder::TreeBuilder<Handle,Sink>::insert_element
  │   37: html5ever::tree_builder::TreeBuilder<Handle,Sink>::step
  │   38: <html5ever::tree_builder::TreeBuilder<Handle,Sink> as html5ever::tokenizer::interface::TokenSink>::process_token
  │   39: html5ever::tokenizer::Tokenizer<Sink>::process_token
  │   40: html5ever::tokenizer::Tokenizer<Sink>::end
  │   41: script::dom::servoparser::ServoParser::do_parse_sync
  │   42: profile_traits::time::profile
  │   43: script::dom::servoparser::ServoParser::parse_sync
  │   44: <script::dom::servoparser::ParserContext as net_traits::FetchResponseListener>::process_response_eof
  │   45: script::script_thread::ScriptThread::handle_new_layout
  │   46: script::script_thread::ScriptThread::process_attach_layout
  │   47: script::dom::htmliframeelement::HTMLIFrameElement::start_new_pipeline
  │   48: <T as script::task::TaskBox>::run_box
  │   49: script::dom::document::Document::remove_script_and_layout_blocker
  │   50: script::dom::node::Node::insert
  │   51: script::dom::node::Node::pre_insert
  │   52: script::dom::bindings::codegen::Bindings::NodeBinding::Node_Binding::appendChild::{{closure}}
  │   53: mozjs::panic::wrap_panic
  │   54: script::dom::bindings::codegen::Bindings::NodeBinding::Node_Binding::appendChild
  │   55: CallJitMethodOp
  │   56: script::dom::bindings::utils::generic_call
  │   57: <unknown>
  │   58: <unknown>
  │   59: <unknown>
  │   60: <unknown>
  │   61: <unknown>
  │   62: <unknown>
  │   63: <unknown>
  │   64: <unknown>
  │   65: <unknown>
  │   66: <unknown>
  │   67: _ZN2js3jit32EnterBaselineInterpreterAtBranchEP9JSContextPNS_16InterpreterFrameEPh
  │   68: <unknown>
  │ mozilla::detail::MutexImpl::~MutexImpl: pthread_mutex_destroy failed: Device or resource busy
  │ Caught signal 11thread panicked while processing panic. aborting.
  │ Redirecting call to abort() to mozalloc_abort
```

</details>

### LINKED ISSUE #32744: tweakers.net servo crash - assertion failed: window.Document ().can invoke script()
**Describe the bug:**
When i visit https://tweakers.net/ in Servo 0.0.1-1c6b74e the page crashes with error:
```
assertion failed: window.Document ().can invoke script()
...
script::dom::bindings::codegen::Bindings::ElementBinding::Element Binding::set_innerTML::{{closure}}
```

**To Reproduce:**
Visit https://tweakers.net/ in the minibrowser

**Platform:**
Servo 0.0.1-1c6b74e
macOS 13.6.7 (22G720) x86

### PR DESCRIPTION
**Pull request by jdm**
These changes adjust the Servo parser implementations to match the interface changes in https://github.com/servo/html5ever/tree/no-mut-interface. Additionally, we can now remove the assertions around active parsers when invoking arbitrary JS (added in #19307) since our parser is now reentrant.

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #32454 and fix #32646 and fix #32744 and fix #24115.
- [x] There are tests for these changes

### CODE REVIEW

**Comment by jdm** [comment]
Woohoo! I've verified that various testcases from the original issue no longer panic, and I have a regression test that panics before the changes and passes after the changes.

File: components/script/dom/element.rs
Comment (jdm) [inline components/script/dom/element.rs:766]: This change appears to be required due to https://github.com/servo/stylo/commit/33c26645a906afbe35821ba567c7c92d96a3ba14, which has not been pulled into Servo since merging into stylo.

**Comment by jdm** [comment]
These changes are now ready for review.

**Comment by sagudev** [comment]
```
error[E0308]: mismatched types
   --> components/fonts/tests/font_context.rs:167:19
    |
167 |             list: names.into_boxed_slice(),
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^ expected `ArcSlice<SingleFontFamily>`, found `Box<[SingleFontFamily]>`
    |
    = note: expected struct `ArcSlice<SingleFontFamily>`
               found struct `std::boxed::Box<[SingleFontFamily]>`
```

File: tests/wpt/mozilla/tests/mozilla/parser-reentrancy-customelement.html
Comment (gterzian) [inline tests/wpt/mozilla/tests/mozilla/parser-reentrancy-customelement.html:3]: Can we get a test with `dom.servoparser.async_html_tokenizer.enabled`? Does the mozilla folder also support adding a `__dir__.ini` in the meta folder? See for example https://github.com/servo/servo/blob/main/tests/wpt/meta/resize-observer/__dir__.ini

File: components/script/dom/servoparser/async_html.rs
Comment (gterzian) [inline components/script/dom/servoparser/async_html.rs:357]: I spent a bit of time looking at the re-entrant safety of this in https://github.com/servo/servo/blob/843726cdf580be6453a0b93f9a3adf0ca071deb2/components/script/dom/servoparser/async_html.rs#L403, and while it seems safe I don't like it. Should we change this to return a domroot? In many cases the dom is turned into one at the point of use anyway. Seems like a risky and probably useless optimization.

**Review by gterzian — APPROVED** [review]
LGTM, with a few questions.

File: components/script/dom/servoparser/async_html.rs
Comment (jdm) [reply components/script/dom/servoparser/async_html.rs:357]: I don't remember the precise history of why we used Dom as the handle type, but since this is preexisting code I'll file a followup issue.

Comment (jdm) [reply components/script/dom/servoparser/async_html.rs:357]: #33011

File: tests/wpt/mozilla/tests/mozilla/parser-reentrancy-customelement.html
Comment (jdm) [reply tests/wpt/mozilla/tests/mozilla/parser-reentrancy-customelement.html:3]: Good call!

Comment (jdm) [reply tests/wpt/mozilla/tests/mozilla/parser-reentrancy-customelement.html]: I've added the new test with the async parser variant, and it's exposed an issue in the async parser that will need more investigation. I propose merging with the current panicking result and doing the investigation in a separate followup.

**Comment by jdm** [comment]
Bah, more changes from https://github.com/servo/stylo/commit/33c26645a906afbe35821ba567c7c92d96a3ba14 that are blocking this PR. I'm going to try to extract those to a separate PR.

**Comment by jdm** [comment]
Once #33079 merges this will (hopefully) be able to finally merge.

**Pull request merged**
